feat: add classes for pseudo metrics with generic codomain

edegeltje opened the pull request:
This PR adds definitions for `GPseudoMetric` and `GPseudoMetricClass`.

`GPseudoMetric a b` is a structure carrying a distance function on the type `a` with a codomain `b`, along with proof that it is a pseudo metric. Furthermore, `GPseudoMetricSpace T a b` declares `T` is a type for pseudo metrics on `a` where the metric takes on values in `b`.
These notions of generic pseudo metric spaces are necessary for at least these two reasons:
1. Some metrics cannot be injectively embedded into Reals or EReals (for example the distance induced by the norm given by absolute value on hyperreal or surreal numbers). 
2. it allows the immediate use of properties of the 'actual' codomain, rather than having to frequently cast between `Real` and your actual codomain. For example, when the actual codomain of the metric is `Nat`, you can define the covering radius in a way which definitionally results in another instance of `Nat`, rather than having to prove it does.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

edegeltje committed:
adding basic definitions for GPseudoMetricSpace
this includes defining:
- IsOrderedAddCommMonoid
- GDist
- GPseudoMetricSpace

edegeltje committed:
add a mixin, add comments
new mixin: IsOrderedCancelAddCommMonoid.
extra comments for GPseudoMetricSpace and others

edegeltje committed:
fix some linter errors

edegeltje committed:
importing added files.
because apparently that's wanted.

edegeltje committed:
simplify imports

edegeltje committed:
change ordered-monoid mixins to be Propositional

edegeltje committed:
now only using the mixins

edegeltje committed:
formatting changes
add and remove empty lines where appropriate

edegeltje committed:
more documentation for the file, formatting
added descriptions of `ball` and `closedBall` at the head of the file
added two newlines between sections and namespaces

edegeltje committed:
add more descriptive comments, formatting
changed descriptions of mixins
added newlines in appropriate places

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+In this file we introduce `GPseudoMetricSpace`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering that
+respects sensible assumptions about interactions between `+` and `≤`.
+## Main Definitions
```
```suggestion
respects sensible assumptions about interactions between `+` and `≤`.

## Main Definitions
```

There should be a blank line between sections and new headings.

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+open scoped BigOperators
+
+/-- Construct a bornology from a generic distance function and metric space axioms.
+this is not declared as an instance, as it would lead to multiple instances of bornology on α,
```
```suggestion
This is not declared as an instance, as it would lead to multiple instances of bornology on α,
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+          (add_le_add le_rfl (ht hy hz)).trans (le_max_right _ _))⟩)
+    fun z => ⟨gdist z z, forall_eq.2 <| forall_eq.2 le_rfl⟩
+
+/-- the class of spaces α with a generic distance function on which return an element of type β -/
```
```suggestion
/-- The class of spaces α with a generic distance function on which return an element of type β -/
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+  /-- the generic distance function on α to β -/
+  protected gdist : α → α → β
+
+/-- the generic distance function on a space α which returns an element of type β,
```
```suggestion
/-- The generic distance function on a space α which returns an element of type β,
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+    _ = gdist x y + gdist x y:= by rw [gdist_comm]
+  exact nonneg_add_self_iff.mp this
+
+/-- generic Pseudo-metric spaces
+a generic Pseudo metric space is a space endowed with some distance function `gdist : α → α → β`
```
```suggestion
/-- Generic Pseudo-metric spaces

A generic Pseudo metric space is a space endowed with some distance function `gdist : α → α → β`
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/Basic.lean:
```diff
@@ -0,0 +1,388 @@
+end
+
+
+namespace GMetricSpace
+section non_cancel
+variable [LinearOrder β] [AddCommMonoid β] [IsOrderedAddCommMonoid β] [GPseudoMetricSpace α β]
```
```suggestion
namespace GMetricSpace

section non_cancel

variable [LinearOrder β] [AddCommMonoid β] [IsOrderedAddCommMonoid β] [GPseudoMetricSpace α β]
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:
```diff
@@ -0,0 +1,48 @@
+/-!
+# IsOrderedAddCommMonoid mixin
+
+this file defines a mixin class for combining kinds orderings and kinds of AddCommMonoid.
```
```suggestion
This file defines a mixin class for combining kinds orderings and kinds of AddCommMonoid.
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:
```diff
@@ -0,0 +1,48 @@
+given `IsOrderedCancelAddCommMonoid α` and `LinearOrder α`
+-/
+
+/--mixin for saying adding a constant does not change weak inequality-/
```
```suggestion
/-- Mixin for saying adding a constant does not change weak inequality-/
```

newell commented on Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:
```diff
@@ -0,0 +1,48 @@
+  /- Adding a constant on the left does not change the order -/
+  protected add_le_add_left : ∀ a b : α, a ≤ b → ∀ c, c + a ≤ c + b
+
+/--mixin for saying removing the adding of a constant does not change weak inequality-/
```
```suggestion
/-- Mixin for saying removing the adding of a constant does not change weak inequality. -/
```

edegeltje committed:
change structure of GPseudoMetricSpace
rather than having a class represent the space with the distance
function, have a structure represent the function itself.

now we have a structure `GPseudoMetric` which
can cast to the function through `FunLike`,
as well as a class `GPseudoMetricClass`, for use in lemmas
that way future things like `GMetric` can benefit from the same lemmas.

edegeltje committed:
fixing comments, fixing possible CI issues

edegeltje committed:
reduce import depth
changes suggested by shake

edegeltje committed:
add capitalisation to comments

edegeltje committed:
more style fixes

edegeltje commented:
it has been pointed out to me that the `IsOrderedAddCommMonoid` already exists in a more general form, namely `CovariantClass`. i'll be refactoring my changes to use that class instead, and remove my mixins.

edegeltje committed:
refactor IsOrderedAddCommMonoid to CovariantClass
and similar such changes

grunweg commented:
Hello! I randomly came across your PR and noticed there is a merge conflict now. Can you merge the master branch, please?

If you do, this PR will also show up in the list of PRs awaiting review :-)

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main Definitions
```
A small comment: I think mathlib doesn't use Capitalisation In Headers much. 
```suggestion
## Main definitions

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
```
I'm wondering if this title can be made more specific. Is there a name for this concept in textbooks or papers? If there is, it could be useful to reference this.

To be clear: if you know this, it's wonderful - but in my personal opinion, this is not required to have your PR merged.

edegeltje replied:
<rant>
I did some digging, and according to [this mathoverflow post](https://mathoverflow.net/a/363831) among others, these are called `[Codomain]-Metrics` or `[Codomain]-valued metrics` (and the associated spaces `[Codomain]-Metric space`, although in my definition the requirements on the codomain is somewhat weaker than described in that post. if you'd like i can change the file to reflect this? also, in the post, both `𝑅` and `Λ` are used to refer to the codomain... since the latter might be confused with conjunction, i would prefer the first, but i have no clue how common either of these uses are... but then again, [this paper](https://arxiv.org/pdf/1502.05002.pdf) does refer to (something similar to) this concept as "generalised metric space". interestingly, all of these references put more conditions on the codomain than i expected... in particular, they expect the ordering to be co- and contravariant compared to addition...
</rant>
tldr: the version in the PR is close to an existing concept, which gets described as "generalised metric space". such spaces are called `R`-Metric spaces or `R`-valued metric spaces where `R` is the codomain, although capital Lambda is also sometimes used. The concept in this PR is slightly weaker with regards to the assumptions on the codomain.

edegeltje replied:
also, according to [this other paper](http://www.tac.mta.ca/tac/reprints/articles/1/tr1.pdf), there are ties with category theory as well

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+
+## Main Definitions
+
+- `GPseudoMetric α β`: A structure containing a distance function on `α` with codomain `β`,
```
Small grammar comment: in English, you continue in lower-case after a colon. Also in a few other places below.

(Some languages are different. In German, for instance, you use upper-case.)
```suggestion
- `GPseudoMetric α β`: a structure containing a distance function on `α` with codomain `β`,

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+
+- `GPseudoMetric α β`: A structure containing a distance function on `α` with codomain `β`,
+which may be equal to 0 for non-equal elements. the distance function is 0 for equal elements,
+is commutative in its arguments, and satisifies the triangle inequality.
```
Two typos: "the distance function"; "satisifies"

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+
+open Set
+
+/-- Generic pseudo metrics
```
You wrote "pseudo-metric" above, but "pseudo metric" here. Please be consistent :-)
(To your consolation, mathlib uses both versions about equally often, so you're in good company.)

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+identical elements, for which the arguments commute, and for which the triangle
+inequality holds. As opposed to a classical pseudo metric, the codomain of this distance
+function is not necessarily ℝ (or ℝ≥0∞), and as a result does not endow α with a uniform
+space.
```
uniform space structure?

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+    (α : Type*) (β : Type*) [LinearOrder β] [AddCommMonoid β]
+    where
+  /-- A distance function on α with values in β -/
+  toFun (x y : α):β
```
```suggestion
  toFun (x y : α) → β

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+  comm' : ∀ x y : α, toFun x y = toFun y x
+  triangle' : ∀ x y z : α, toFun x z ≤ toFun x y + toFun y z
+
+variable {α β :Type*} [LinearOrder β] [AddCommMonoid β] [CovariantClass β β (.+.) (.≤.)]
```
Here and elsewhere: please put spaces around the colon, and around binary operators (i.e., `+` and `≤`).

And a blank line after the `variable`, please.

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+variable {α β :Type*} [LinearOrder β] [AddCommMonoid β] [CovariantClass β β (.+.) (.≤.)]
+@[ext]
+theorem GPseudoMetric.ext {d₁ d₂ : GPseudoMetric α β} (h : d₁.toFun = d₂.toFun) : d₁ = d₂ := by
+  cases' d₁ with d _ _ _ B hB
```
Often, for such proofs it suffices to write just
```
cases d₁
cases d₂
congr```
Does this also suffice here?

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+  comm' := GPseudoMetric.comm'
+  triangle' := GPseudoMetric.triangle'
+
+
+variable {T:Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist:T)
+
+
```
As once above, please add spaces around the colon - and one blank line around them is enough :-)
```suggestion
variable {T : Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist : T)

grunweg commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,328 @@
+theorem triangle_right (x y z : α) : gdist x y ≤ gdist x z + gdist y z := by
+  rw [comm' gdist y]; apply triangle'
+
+
```
One blank line between these theorems is enough :-)

grunweg reviewed (commented):
Thank you for your contribution! I gave this a superficial look; as a not-really-expert on this area of mathlib, the file looks in broadly right shape, but a more expert reviewer will have to judge the details.

I have a number of small comments: that is perfectly normal at this point (and I like paying attention to details). Just in case: have you seen the [style guide](https://leanprover-community.github.io/contribute/style.html) already? Some things I'm saying are also repeated there, in case you want to take a look later.

edegeltje committed:
Merge branch 'master' into blizzard_inc/gmetric

edegeltje committed:
fix import ordering

Commit:
Update Mathlib.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

edegeltje committed:
"all files need to be imported" fix (?)

edegeltje committed:
apply feedback

edegeltje committed:
fix namespace stuff, comment formatting

edegeltje committed:
mor fix

edegeltje committed:
even mor fix

YaelDillies commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,332 @@
+
+Additional useful definitions:
+
+- `ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
+- `closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
+- `sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```
Please fully qualify names so that the links work in the docs
```suggestion
- `GMetric.ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
- `GMetric.closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
- `GMetric.sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```

YaelDillies commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
```diff
@@ -0,0 +1,332 @@
+As opposed to a classical pseudo-metric, the codomain of this distance function is not
+necessarily `ℝ` (or `ℝ≥0∞`), and as a result does not endow `α` with a uniform space structure.
+-/
+
```
```suggestion
```

YaelDillies commented on Mathlib/Topology/GPseudoMetric/Basic.lean:
You are not talking about pseudo-metrics only, so you might as well call your file `Topology.GMetric.Basic` or `Topology.GMetric`

YaelDillies reviewed (commented):
Is there a specific reason why you are using the "type of metrics" + "hom class" design (which we use for morphisms in algebra, eg `MonoidHom` + `MonoidHomClass`) rather than the "canonical metric" design (`PseudoMetricSpace` or `MetricSpace`).

If so, can you underline that reason in the module doc. If not, can you drop that approach and use the canonical metric approach instead? (please let me know if you don't know what needs to be done to switch)

riccardobrasca commented:
@edegeltje Are you still working on this PR? I think there are design decisions we are still not sure about, see [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Packings.2C.20coverings.2C.20nets.2C.20uniform.20discretesness.20et.2E.20al.2E). Maybe you can discuss the PR there? Thanks!

edegeltje commented:
i have for now abandoned working on this. although i do believe creating this is the correct direction compared to the current class-based unbundled approach, i don't know how to make progress with this...

edegeltje committed:
Merge branch 'master' into blizzard_inc/gmetric

edegeltje committed:
fix merge

semorrison commented:
Discussion amongst reviewers seems to have concluded that a design more like
```
class NatDist (α : Type*) [PseudoEMetricSpace α] where
  natDist : α → α → ℕ
  natCast_natDist (a b) : natDist a b = edist a b
```
is the right path forward here.

Given your comment above, I'm going to close this for now, but hopefully we will see further progress on this when you have time again!
